modules/placement-driver/src/main/java/org/apache/ignite/internal/placementdriver/leases/LeaseTracker.java

```diff
@@ -230,6 +239,49 @@ public void onError(Throwable e) {
         }
     }
 
+    /**
+     * Wait for a new primary replica in case when the current one is inoperable.
+     *
+     * @param groupId Group id.
+     * @param timestamp CLOCK_SKEW aware timestamp reference value.
+     * @param inoperableLeaseStartTime Start time of inoperable lease.
+     * @param timeout How long to wait before completing exceptionally with a TimeoutException, in milliseconds.
+     * @return New leaseholder future.
+     */
+    private CompletableFuture<ReplicaMeta> awaitNewPrimaryReplica(
+            ReplicationGroupId groupId,
+            HybridTimestamp timestamp,
+            HybridTimestamp inoperableLeaseStartTime,
+            long timeout
+    ) {
+        var resultFuture = new CompletableFuture<ReplicaMeta>()
+                .orTimeout(timeout, TimeUnit.MILLISECONDS);
+
+        EventListener<PrimaryReplicaEventParameters> listener = (evt, e) -> {
+            if (evt.groupId().equals(groupId) && !evt.startTime().equals(inoperableLeaseStartTime)) {
+                resultFuture.complete(new Lease(evt.leaseholder(), evt.leaseholderId(), evt.startTime(), evt.expirationTime(), groupId));
+
+                return trueCompletedFuture();
+            } else {
+                return falseCompletedFuture();
+            }
+        };
+
+        listen(PRIMARY_REPLICA_ELECTED, listener);
+
+        CompletableFuture<ReplicaMeta> replicaMetaFut = awaitPrimaryReplica(groupId, timestamp, timeout, TimeUnit.MILLISECONDS);
+
+        replicaMetaFut.thenAccept(replicaMeta -> {
+            if (!replicaMeta.getStartTime().equals(inoperableLeaseStartTime)) {
+                resultFuture.complete(replicaMeta);
+
+                removeListener(PRIMARY_REPLICA_ELECTED, listener);
+            }
+        });
+
+        return resultFuture;
+    }
+
     @Override
     public CompletableFuture<ReplicaMeta> awaitPrimaryReplica(
```

Comment:
Despite the fact that there's an enabled test I'd also add unit ones that will verify the invariants.

Reply:
added test to PlacementDriverTest